File: frontend/src/types/workflow.ts
export type NodeType =
  | 'manual_trigger'
  | 'webhook_trigger'
  | 'transform_data'
  | 'http_request'
  | 'wait'
  | 'decision'
  | 'end'

export type TransformationType =
  | 'uppercase'
  | 'append_text'
  | 'prepend_text'
  | 'multiply'
  | 'rename_key'

export interface NodePosition {
  x: number
  y: number
}

export interface NodeConfig {
  id: string
  type: NodeType
  config: Record<string, unknown>
  position: NodePosition
  label?: string
}

export interface EdgeConfig {
  id: string
  source_node_id: string
  target_node_id: string
  source_handle?: string | null
}

export interface WorkflowDefinition {
  id: string
  name: string
  nodes: NodeConfig[]
  edges: EdgeConfig[]
  created_at?: string
  updated_at?: string
}

export interface ExecutionLog {
  step: number
  node_id: string
  node_type: string
  input_payload: Record<string, unknown>
  output_payload: Record<string, unknown>
  message: string
  error?: string | null
}

export interface WorkflowResult {
  workflow_id: string
  run_id: string
  status: 'completed' | 'failed'
  logs: ExecutionLog[]
  final_output?: Record<string, unknown> | null
  error?: string | null
}

File: backend/app/models/workflow.py
"""
Pydantic models for workflow definitions.

A workflow is a DAG of nodes connected by edges.
- NodeConfig: describes a single node (type, id, config, position)
- EdgeConfig: a directed connection from one node to another
- WorkflowDefinition: the full graph (nodes + edges + metadata)
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import uuid4

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Node types — extend this Literal when adding new node types
# ---------------------------------------------------------------------------
NodeType = Literal[
    "manual_trigger",
    "webhook_trigger",
    "transform_data",
    "http_request",
    "send_email",
    "wait",
    "decision",
    "end",
]


class Position(BaseModel):
    x: float = 0.0
    y: float = 0.0


class NodeConfig(BaseModel):
    """Represents a single node in the workflow graph."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    type: NodeType
    # Node-specific configuration dict (validated per-node at execution time)
    config: dict[str, Any] = Field(default_factory=dict)
    # Canvas position — stored but not used by the engine
    position: Position = Field(default_factory=Position)
    label: Optional[str] = None


class EdgeConfig(BaseModel):
    """A directed edge from source_node → target_node."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    source_node_id: str
    target_node_id: str
    # For Decision nodes: which output handle this edge originates from
    # e.g. "true" | "false" | None (for non-branching nodes)
    source_handle: Optional[str] = None


class WorkflowDefinition(BaseModel):
    """Complete workflow: nodes + edges + metadata."""

    id: str = Field(default_factory=lambda: str(uuid4()))
    name: str = "Untitled Workflow"
    nodes: list[NodeConfig] = Field(default_factory=list)
    edges: list[EdgeConfig] = Field(default_factory=list)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


# ---------------------------------------------------------------------------
# API request / response models
# ---------------------------------------------------------------------------

class RunWorkflowRequest(BaseModel):
    """Sent by the client to trigger a workflow run."""
    workflow_id: str
    # Optionally override the trigger payload inline
    trigger_payload: dict[str, Any] = Field(default_factory=dict)


class RunWorkflowResponse(BaseModel):
    workflow_id: str
    run_id: str
    status: str = "started"


class ExecutionLog(BaseModel):
    step: int
    node_id: str
    node_type: str
    input_payload: dict[str, Any]
    output_payload: dict[str, Any]
    message: str = ""
    error: Optional[str] = None


class WorkflowResult(BaseModel):
    workflow_id: str
    run_id: str
    status: Literal["completed", "failed"]
    logs: list[ExecutionLog]
    final_output: Optional[dict[str, Any]] = None
    error: Optional[str] = None
